lxd/devlxd.go

```diff
@@ -118,21 +118,43 @@ var devlxdEventsGet = devLxdHandler{"/1.0/events", func(d *Daemon, c instance.In
 		typeStr = "config,device"
 	}
 
-	conn, err := shared.WebsocketUpgrader.Upgrade(w, r, nil)
+	if r.Header.Get("Upgrade") == "websocket" {
+		conn, err := shared.WebsocketUpgrader.Upgrade(w, r, nil)
+		if err != nil {
+			return &devLxdResponse{"internal server error", http.StatusInternalServerError, "raw"}
+		}
+		defer conn.Close() // This ensures the go routine below is ended when this function ends.
+
+		listener, err := d.devlxdEvents.AddListener(c.ID(), conn, strings.Split(typeStr, ","))
+		if err != nil {
+			return &devLxdResponse{"internal server error", http.StatusInternalServerError, "raw"}
+		}
+
+		logger.Debug("New container event listener", logger.Ctx{"instance": project.Instance(c.Project(), c.Name()), "listener_id": listener.ID})
```

Comment:
Can we use separate fields for `instance` and `project` here rather than joining them together.